Add djvu layer

dalanicolai opened the pull request:
This PR adds a djvu layer. Additional to the [djvu](https://elpa.gnu.org/packages/djvu.html) package, the layer installs and configures the [djvu3](https://github.com/dalanicolai/djvu3) (extension) package that adds many useful functionalities like rendering of annotations, a fast search and a djvu-occur search function, imenu navigation and more.

lebensterben commented on layers/+readers/djvu/packages.el:
```diff
@@ -0,0 +1,107 @@
+;;; packages.el --- djvu layer packages file for Spacemacs.
+;;
+;; Copyright (c) 2012-2018 Sylvain Benner & Contributors
```
```suggestion
;; Copyright (c) 2021 Sylvain Benner & Contributors
```

lebensterben commented on layers/+readers/djvu/README.org:
```diff
@@ -0,0 +1,86 @@
+| ~SPC j i~   | imenu navigation                                                                 |
+| ~o~         | outline                                                                          |
+
+/Annotations/
```
```suggestion
*Annotations*
```

lebensterben commented on layers/+readers/djvu/funcs.el:
```diff
@@ -0,0 +1,71 @@
+(defun djvu-toggle-semi-continuous-scrolling ()
+  (interactive)
+  (setq djvu-semi-continuous-scrolling (if djvu-semi-continuous-scrolling
```
```elisp
(setq djvu-semi-continuous-scrolling (not djvu-semi-continuous-scrolling))
```

lebensterben commented on layers/+readers/djvu/funcs.el:
```diff
@@ -0,0 +1,71 @@
+                                               "enabled"
+                                             "disabled")))
+
+(defun spacemacs/djvu-advise-image-toggle (orig-func &rest args)
```
```suggestion
(defun spacemacs/djvu-advise-image-toggle (_file &rest _args)
```

The first argument is `file`.
Since they are not used, they should be prefixed with underscore.

See https://github.com/dalanicolai/djvu2.el/blob/c89f513e589e06bcf239db0f502ed9bac34753d1/djvu.el#L1732

lebensterben commented on layers/+readers/djvu/funcs.el:
```diff
@@ -0,0 +1,71 @@
+      (if djvu-image-mode
+          (djvu-image-scroll-down)
+        (evil-previous-visual-line))
+    (if (not (= (window-vscroll) 0))
```
```suggestion
    (if (/= (window-vscroll) 0)
```

lebensterben commented on layers/+readers/djvu/funcs.el:
```diff
@@ -0,0 +1,71 @@
+
+(defun spacemacs/djvu-search-forward (query)
+  "Search forward for match for REGEXP.
+
```
No need to add a blank line after first line of docstring.
It's added automatically.

lebensterben commented on layers/+readers/djvu/funcs.el:
```diff
@@ -0,0 +1,71 @@
+  "Search forward for match for REGEXP.
+
+Search case-sensitivity is determined by the value of the variable
+`case-fold-search', which see.
```
>  which see.

is this sentence complete?

lebensterben commented on layers/+readers/djvu/funcs.el:
```diff
@@ -0,0 +1,71 @@
+                  (eq (djvu-ref page) (djvu-ref pagemax))))
+    (djvu-goto-page (let ((page (djvu-ref page))
+                          (return 1))
+                      (while (and (not (= return 0)) (< page (+ (djvu-ref pagemax) 1)))
```
```suggestion
                      (while (and (/= return 0) (< page (+ (djvu-ref pagemax) 1)))
```

lebensterben commented on layers/+readers/djvu/img/djvu-logo.svg:
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
what's the license of this file

dalanicolai replied:
Well, I can add the source of the image (I guess as a caption?), but don't know about any explicit permission.

lebensterben commented on layers/+readers/djvu/packages.el:
```diff
@@ -0,0 +1,107 @@
+    :init
+    (add-to-list 'spacemacs-large-file-modes-list 'djvu-read-mode t)
+    (add-hook 'djvu-read-mode-hook (lambda () (setq imenu-create-index-function #'djvu-imenu-create-index)))
+    (add-hook 'djvu-read-mode-hook (lambda () (setq imenu-default-goto-function (lambda (title page) (djvu-goto-page page djvu-doc)))))
```
```suggestion
    (add-hook 'djvu-read-mode-hook
              (lambda ()
                (setq imenu-default-goto-function (lambda (title page)
                                                    (djvu-goto-page page djvu-doc)))))
```

lebensterben commented on layers/+readers/djvu/packages.el:
```diff
@@ -0,0 +1,107 @@
+      "bb" 'djvu-switch-bookmarks)
+
+    ;; for some reason can not use dolist here
+      (define-key djvu-read-mode-map [remap save-buffer] 'djvu-save)
```
these three line doesn't seem to be indented correctly, 2 extra spaces in the beginning of line.

lebensterben commented on layers/+readers/djvu/README.org:
```diff
@@ -0,0 +1,87 @@
+| ~i~         | djvu-image-toggle                                                                |
+| ~j~         | scroll up                                                                        |
+| ~k~         | scroll down                                                                      |
+| ~J~         | next page                                                                        |
```
that's okay.
but can you also include `h/l` for horizontal scrolling in the table?

then if `H/L` are not occupied, could it be used for command like previous/next chapter/section or bookmarked page?

dalanicolai replied:
The djvu package does not provide horizontal scrolling commands. But they are not very needed, because when zooming in, the package stops rendering images already before the page width reaches half the(my) screensize...

lebensterben commented on layers/+readers/djvu/README.org:
```diff
@@ -0,0 +1,87 @@
+| ~+/-~       | zoom-in/out                                                                      |
+| ~, s~       | djvu-occur                                                                       |
+| ~/~         | djvu fast search                                                                 |
+| ~n~         | djvu search continue                                                             |
```
https://github.com/dalanicolai/djvu2.el/blob/c89f513e589e06bcf239db0f502ed9bac34753d1/djvu.el#L2198-L2210

this underlying implementation can be improved by keeping a counter `i`, and when the loop end, `(djvu-next-page i)`.

dalanicolai replied:
I am not using the file djvu2.el in this PR. This PR contains djvu.el (from ELPA) with quite some extensions in djvu3.el and some improved functions in funcs.el. I have tried to get the improvements into djvu.el but there was neither a response to my message to the [bug-gnu-emacs](https://lists.gnu.org/archive/html/bug-gnu-emacs/2021-05/msg01755.html) nor to my message to the [emacs-devel](https://lists.gnu.org/archive/html/emacs-devel/2021-05/msg00944.html) mailing list.

Anyway, I haven't bound the function you are pointing to (the original function from djvu.el), because I have added a faster simple search version in `funcs.el`. But as I mentioned, I did not bother to improve it further because users really should use the `djvu-occur` function, if they prefer even better search functionality.

lebensterben replied:
well, except a few limited cases, a symbol in package `foo` should be prefixed with `foo-`.
so you need to rename the symbols in `djvu3`, or otherwise it's really confusing.

dalanicolai replied:
Ah okay, well I had named them djvu because they were destined to get merged into djvu.el. Also, I am overriding functions of djvu.el, so for that I should really use the same name of the function in djvu.el (or create an alias), because these functions are used in djvu.el itself. Anyway, if the PR gets merged, then when users jump to the definition of the functions, they will end up in djvu3.el, so I think it is not too confusing (although it is a little unfortunately). Anyway, `djvu2.el` was not involved in this PR at all.

dalanicolai commented:
I got confused by cd0c943. It seems that when I accept your suggested changes here on github, an extra commit is created, which is not preferable of course. So I have reverted to a single commit and rebased on develop.

lebensterben commented:
please check the remaining unresolved request.

lebensterben commented on layers/+readers/djvu/config.el:
```diff
@@ -0,0 +1,24 @@
+;;; config.el --- Djvu Layer Configuration File for Spacemacs
+;;
+;; Copyright (c) 2012-2021 Sylvain Benner & Contributors
```
```suggestion
;; Copyright (c) 2021 Sylvain Benner & Contributors
```

smile13241324 commented:
I have merged the PR and resolved the conflict in changelog.develop. I have also revised the layer and fixed the prefixes as well as the keybinding tables in the  README. I have also removed the empty layer file for I did not see why a specific dependency on `spacemacs-evil` would be necessary.

@dalanicolai care to double check?

dalanicolai commented:
Thanks Maxi! I have looked at it and it all looks good to me. I don't remember why I added that dependency on `spacemacs-evil`, but indeed I do not see directly why that would be required. Great to have a nice djvu reader/editor in Spacemacs now!
